spec/index.html

```diff
@@ -1492,3 +1502,366 @@ <h1>SharedArray ( ..._values_ )</h1>
+        </emu-alg>
+      </emu-clause>
+
+      <emu-clause id="sec-atomics.mutex.unlocktoken.prototype.unlock">
+        <h1>Atomics.Mutex.UnlockToken.prototype.unlock ( )</h1>
+        <p>This function performs the following steps when called:</p>
+        <emu-alg>
+          1. Let _token_ be the *this* value.
+          1. Perform ? RequireInternalSlot(_token_, [[LockedMutex]]).
+          1. Let _mutex_ be _token_.[[LockedMutex]].
+          1. If _mutex_ is not ~empty~, then
+            1. Set _token_.[[LockedMutex]] to ~empty~.
+            1. Perform UnlockMutex(_mutex_).
+            1. Return *true*.
+          1. Return *false*.
+        </emu-alg>
+      </emu-clause>
+
+      <emu-clause id="sec-atomics.mutex.unlocktoken.prototype.%symbol.dispose%">
+        <h1>Atomics.Mutex.UnlockToken.prototype [ %Symbol.dispose% ] ( )</h1>
+        <p>The initial value of the %Symbol.dispose% property is %Atomics.Mutex.UnlockToken.prototype.unlock%, defined in <emu-xref href="#sec-atomics.mutex.unlocktoken.prototype.unlock"></emu-xref>.</p>
+      </emu-clause>
+    </emu-clause>
+  </emu-clause>
+
+  <emu-clause id="sec-condition-object">
+    <h1>The Condition Constructor</h1>
+    <p>The Condition constructor:</p>
+    <ul>
+      <li>is <dfn>%Atomics.Condition%</dfn>.</li>
+      <li>is the initial value of the *"Condition"* property of the %Atomics% object.</li>
+      <li>creates and initializes a new Condition when called as constructor.</li>
+      <li>is not intended to be called as a function and will throw an exception when called in that manner.</li>
+    </ul>
+
+    <emu-clause id="sec-atomics.condition">
+      <h1>Atomics.Condition ( )</h1>
+      <p>This function performs the following steps when called:</p>
+      <emu-alg>
+        1. If NewTarget is *undefined*, throw a *TypeError* exception.
+        1. Let _createInitializer_ be a new Abstract Closure with parameters (_newCV_) that captures nothing and performs the following steps when called:
+          1. Set _newCV_.[[ConditionWaiterList]] to a new WaiterList Record.
+          1. Return ~unused~.
+        1. Return SharedStructCreate(_createInitializer_, « [[ConditionWaiterList]] »).
+      </emu-alg>
+    </emu-clause>
+  </emu-clause>
+
+  <emu-clause id="sec-properties-of-the-condition-constructor">
+    <h1>Properties of the Condition Constructor</h1>
+    <p>The Condition constructor:</p>
+    <ul>
+      <li>has a [[Prototype]] internal slot whose value is %Function.prototype%.</li>
+      <li>has the following properties:</li>
+    </ul>
+
+    <emu-note>
+      <p>Per-Realm prototypes, which is currently an open design question and not included in this draft, will give Conditions a per-Realm prototype with built-in methods instead of static methods.</p>
+    </emu-note>
+
+    <emu-clause id="sec-atomics.condition.wait">
+      <h1>Atomics.Condition.wait ( _cv_, _mutexUnlockToken_ [ , _timeout_ ] )</h1>
```

Comment:
What is the chance we can add a mechanism to combine `wait` with a callback? It's a fairly convenient way to combine waiting with actually checking a condition.

In [`@esfx/threading-conditionvariable`](https://esfx.js.org/esfx/api/threading-conditionvariable/conditionvariable.html#_esfx_threading_conditionvariable_ConditionVariable_class) I use two separate methods:
- `ConditionVariable.prototype.wait(mut, callback?)`
- `ConditionVariable.prototype.waitFor(mut, timeoutMs, callback?)`

Since its common to combine a condition variable with an actual condition, it's far more convenient to be able to write:

```js
{
  using lck = Atomics.Mutex.lock(mut);
  Atomics.Condition.wait(cv, lck, () => data.processed);
  ...
}
```
than it is to write the following:
```js
{
  using lck = Atomics.Mutex.lock(mut);
  while (!data.processed) {
    Atomics.Condition.wait(cv, lck);
  }
  ...
}
```

And **far** more convenient if you're combining a callback with a timeout, especially given the edge cases:

```js
{
  using lck = Atomics.Mutex.lock(mut);
  if (Atomics.Condition.waitFor(cv, lck, timeoutMs, () => data.processed) === "timed-out") {
    // handle timeout
  }
  else {
    // handle ok
  }
}
```
vs
```js
{
  using lck = Atomics.Mutex.lock(mut);
  const absoluteTimeout = Date.now() + timeoutMs;
  let result = "ok";
  while (result === "ok" && !data.processed) {
    const remainingTimeout = absoluteTimeout - Date.now();
    result = Atomics.Condition.wait(cv, lck, remainingTimeout);
  }
  if (result === "timed-out") {
    // handle timeout
  }
  else {
    // handle ok
  }
  ...
}
```

Reply:
To make sure I understand, the callback would only be called in the case of being notified, and its return value would be ignored?

Reply:
The callback is invoked when notified after the `Condition` has reacquired the lock. If the return value is false (or falsy), the `Condition` releases the lock and starts waiting again, similar to overload (2) in https://en.cppreference.com/w/cpp/thread/condition_variable/wait.

Naively, this can be modeled in user code as:
```js
const wait = (cond, tok, pred) => {
  while (!pred()) {
    Atomics.Condition.wait(cond, tok);
  }
};
```

However, modeling `waitFor` with a timeout is more complex as you need to track how much of the timeout has elapsed since you started waiting, not since the last invocation of `pred`.

Reply:
I do think a `pred`-taking overload is useful like in the C++ standard library for the same reasons you say. But there is a difference in utility: C++ allows spurious wakeups, while we don't (as in, we currently don't for `Atomics.wait` and I see no reason to relax this for `Atomics.Condition`).

A predicate-taking overload is still useful (e.g., a workstealing queue where several workers could all be notified, but some might have no work), but we don't need it for as many cases as C++ does.

Reply:
I've split `Condition.wait` into a blocking-forever variant, `wait`, and a timeout and predicate-taking variant, `waitFor`. This also works around the return type oddness.

- `wait` always returns `undefined`.
- `waitFor`, if passed a predicate, always returns the value of the final call to the predicate. If not passed a predicate, returns `true` for having been notified and `false` for timing out.

Reply:
We _could_ just have `waitFor` (and rename it back to `wait`), but I kind of like a simpler, common-case API having a different name instead of one overload for everything.

Reply:
So, if I do have a case where I need to wait indefinitely but check a condition, I would use `waitFor(cv, token, Infinity, pred)`?

Reply:
> So, if I do have a case where I need to wait indefinitely but check a condition, I would use waitFor(cv, token, Infinity, pred)?

Yes.